### PR TITLE
feat(booking): user edit own booking endpoint

### LINKED ISSUE #690: [BACKEND] Create user endpoint to edit their own bookings
**Is your feature request related to a problem? Please describe.**

Need to create `PATCH /api/me/bookings/[id]` route for users to edit their **own** bookings. This indicates that there should be certain logic to check the request payload to ensure that the affected booking belongs to the authenticated user. 

You should also check the number of game session slots available and return an appropriate error code if there is an error this way. 

Feel free to ask if any data-layer services are missing and to add them when needed. 

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**BEFORE MERGING**
- [ ] Integration test written for services
- [ ] [Schemas annotated](https://payloadcms.com/docs/configuration/collections) if adding new models
- [ ] Code generation run (*hint*: `pnpm generate:types`)
- [ ] Appropriate mocks created where possible
- [ ] PR Reviewed (For non-trivial changes)
- [ ] Changes tested after rebasing on master or merging in master (*hint*: `git fetch origin master:master`, then `git rebase master` or `git merge master`)
- [ ] All required PR checks passing

### PR DESCRIPTION
**Pull request by DerrickTrang1**
# Description

> [!NOTE]
> We don't check if the user is the same within the object as the schema omits it

This route handles all of the user modifications towards their own bookings. The route logic covers the following:

1. Checks if a booking exists
2. Checks if the user owns that booking (returns 404 otherwise)
3. Checks if the booking start time has not already passed (this ensures that a user doesn't continuously modify a single booking for an "infinite" loop)
4. Checks if a game session modification was made (by checking if it was specified in the object and compares with the existing booking data)
  a. No change has made -> Modifies the booking object and commits
  b. The following logic checks occur:
    1. Check if the new session exists (returns 404 with different message to booking not found)
    2. Checks the session open time, if this is in the future, return 403
    3. Checks if a booking has already been made to this target session (returns 409 if dupe booking)
    4. Checks all existing bookings for the new session and ensures that the limit hasn't been reached (it also checks the user role to determine which capacity type to check)
6. Modifies the booking object and commits

Extensive tests have also been added to cover all of the scenarios above (please double check when reviewing to ensure edge cases are covered). 

Fixes #690 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [x] Integration tests written (requires checks to pass)

## Checklist before requesting a review

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added thorough tests that prove my fix is effective and that my feature works
- [x] I've requested a review from another user

### CODE REVIEW

File: apps/backend/src/app/api/me/bookings/[id]/route.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        typeof parsedBody.gameSession === "string"
+          ? await gameSessionDataService.getGameSessionById(parsedBody.gameSession)
+          : parsedBody.gameSession
+
+      if (!gameSession) {
+        return NextResponse.json(
+          { error: "Game session not found" },
+          { status: StatusCodes.NOT_FOUND },
+        )
+      }
```
Comment (jeffplays2005) [inline apps/backend/src/app/api/me/bookings/[id]/route.ts:31]: if `gameSessionDataService.getGameSessionById` fails, it returns a `NotFound` exception

File: apps/backend/src/app/api/me/bookings/[id]/route.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        )
+
+      if (
+        (await bookingDataService.getAllUserBookingsBySessionId(userData.id, gameSession.id))
+          .length > 0
+      )
+        return NextResponse.json(
+          { error: "Session already booked" },
+          { status: StatusCodes.CONFLICT },
+        )
```
Comment (jeffplays2005) [inline apps/backend/src/app/api/me/bookings/[id]/route.ts:63]: logically better to show conflict before max number of sessions reached

i.e. if a user already books something but tries to update their booking, ideally, they see that they already have a booking rather than `"Session is full for the selected user role"`

File: apps/backend/src/app/api/me/bookings/[id]/route.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import GameSessionDataService from "@/data-layer/services/GameSessionDataService"
+import UserDataService from "@/data-layer/services/UserDataService"
+
+class RouteWrapper {
+  @Security("jwt")
+  static async PATCH(req: RequestWithUser, { params }: { params: Promise<{ id: string }> }) {
+    const { id } = await params
+    const gameSessionDataService = new GameSessionDataService()
+    const userDataService = new UserDataService()
+    const bookingDataService = new BookingDataService()
```
Comment (jeffplays2005) [inline apps/backend/src/app/api/me/bookings/[id]/route.ts:16]: logic is missing two crucial check

1. if a booking has already passed, cant the user just modify their expired booking to a current one and get free sessions forever?
2. there isn't an initial check to see if the booking even exists

Comment (jeffplays2005) [reply apps/backend/src/app/api/me/bookings/[id]/route.ts:16]: - [x] check if booking exist
- [x] check booking date past

File: apps/backend/src/app/api/me/bookings/[id]/route.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        }
+
+        const bookings = await bookingDataService.getAllBookingsBySessionId(gameSession.id)
+        // Refetch user data as JWT stored data could be outdated
+        const userData = await userDataService.getUserById(req.user.id)
+        if (bookings.some((booking) => getPayloadObjectId(booking.user) === userData.id))
+          return NextResponse.json(
+            { error: "A booking with that session has already been made" },
+            { status: StatusCodes.CONFLICT },
+          )
```
[the remark this answers is not in the document]
Comment (jeffplays2005) [reply apps/backend/src/app/api/me/bookings/[id]/route.ts:79]: There's already a check to ensure that the new game session isn't the same as the old one on line 44

File: apps/backend/src/app/api/me/bookings/[id]/route.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+        if (
+          (userData.role === MembershipType.casual &&
+            bookings.length >= gameSession.casualCapacity) ||
+          (userData.role === MembershipType.member && bookings.length >= gameSession.capacity)
```
Comment (choden-dev) [inline apps/backend/src/app/api/me/bookings/[id]/route.ts:86]: nit: switch may be clearer here as for what you're trying to do

File: apps/backend/src/app/api/me/bookings/[id]/route.ts [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    try {
+      const initialBooking = await bookingDataService.getBookingById(id)
+      // Throw payload not found when the user doesn't own this booking
+      if (getPayloadObjectId(initialBooking.user) !== req.user.id) {
+        throw new NotFound()
+      }
+
+      const now = new Date()
+
+      if (new Date((initialBooking.gameSession as GameSession).startTime) < now) {
```
Comment (choden-dev) [inline apps/backend/src/app/api/me/bookings/[id]/route.ts:34]: question: how is the `initialBooking` object not typesafe?

**Review by choden-dev — APPROVED** [review]
he
